src/resolve/resolved-dns-stream.c

```diff
@@ -441,6 +453,34 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
                         log_warning_errno(errno, "Couldn't restart TCP connection timeout, ignoring: %m");
         }
 
+#if ENABLE_DNS_OVER_TLS
+        if (s->encrypted) {
+                uint32_t events;
+
+                r = sd_event_source_get_io_events(s->io_event_source, &events);
+                if (r < 0)
+                        return r;
+
+                /* When using DNS-over-TLS, the underlying TLS library may read the entire TLS record
+                   and buffer it internally. If this happens, we will not receive further EPOLLIN events,
+                   and unless there's some unrelated activity on the socket, we will hang until time out.
+                   To avoid this, if there's buffered TLS data, generate a "fake" EPOLLIN event.
+                   This is hacky, but it makes this case transparent to the rest of the IO code. */
+                if ((events & EPOLLIN) != 0 && dnstls_stream_has_buffered_data(s)) {
+                        if (!s->tls_buffered_data_event_source) {
+                                r = sd_event_add_defer(s->manager->event, &s->tls_buffered_data_event_source,
```

Comment:
Why do not you simply loop this function?

Reply:
Looping is possible, it's more of an style thing:
(1) Besides the declarations the entire handling of the edge case is contained in a single block
(2) Looping/recursing/goto'ing to the top of the function feels ugly (more indentation or stack frames or labels, inside #ifdefs for ENABLE_DNS_OVER_TLS)
(3) The code flow is more unified with the plain TCP case, e.g. an event emits at most one DNS packet

But I understand that emitting a "fake IO" event is also a bit ugly and less performant, so if you feel looping is preferable, I can change that.

Reply:
If looping also works, then I prefer to convert the logic you added into loop.

How about to rename `on_stream_io()` to e.g. `on_stream_io_impl()` and then
```
static int on_stream_io_impl(DnsStream *s, uint32_t revents) {
        ...
}

static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *userdata) {
        _cleanup_(dns_stream_unrefp) DnsStream *s = NULL;
        int r;

        assert(userdata);

        s = dns_stream_ref(userdata); /* Protect stream while we process it */

        r = on_stream_io(s, revents);
        if (r < 0)
                return r;

#if ENABLE_DNS_OVER_TLS
        if (!s->encrypted)
                return 0;
        if (!FLAGS_SET(events, EPOLLIN))
                return 0;

        while (dnstls_stream_has_buffered_data(s)) {
                /* When using DNS-over-TLS, the underlying TLS library may read the entire TLS record
                   and buffer it internally. If this happens, we will not receive further EPOLLIN events,
                   and unless there's some unrelated activity on the socket, we will hang until time out.
                   To avoid this, ....... */
                r = on_stream_io_impl(s, revents);
                if (r < 0)
                        return r;
        }
#endif

        return 0;
}
```